# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: ** This is a Classification problem. As the model must rightly predict students passing or failing.

From the project description itself:
"As a software engineer contacted by the school district, your task is to model the factors that predict how likely a student is to pass their high school final exam, by constructing an intervention system that leverages supervised learning techniques."

Using this as an indicator of the direction for the project, it seems to be a classification problem with focus on students passing. It seems that a good model would be one that has a high metric based on the passing students.

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [4]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [5]:
# TODO: Calculate number of students
n_students = len(student_data)

# TODO: Calculate number of features
n_features = len(student_data.columns)

# TODO: Calculate passing students
n_passed = len(student_data[student_data['passed'] == 'yes'])

# TODO: Calculate failing students
n_failed = len(student_data[student_data['passed'] == 'no'])

# TODO: Calculate graduation rate
grad_rate = n_passed * 100 / n_students

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 31
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.00%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [6]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [7]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [8]:
# TODO: Import any additional functionality you may need here
from sklearn.cross_validation import train_test_split

# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train

# TODO: Shuffle and split the dataset into the number of training and testing points above
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size = 0.24, random_state = 1)

# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])
print "Yes in Train: {}".format((y_train == 'yes').sum())
print "No in Train: {}".format((y_train == 'no').sum())
print "Yes in Test: {}".format((y_test == 'yes').sum())
print "No in Test: {}".format((y_test == 'no').sum())

Training set has 300 samples.
Testing set has 95 samples.
Yes in Train: 198
No in Train: 102
Yes in Test: 67
No in Test: 28


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

**Answer: **

Three Supervised Learning models appropriate for this problem are:

1. Support Vector Machines(SVM)
2. K-Nearest Neighbors(KNeighbors)
3. Logistic Regression

**1. Support Vector Machines**

Real World Application:

General Application: Image Clustering, Texture Classification, Text Classification, Disease Detection etc.

Support Vector Machines were used for detecting persons with common diseases such as diabetes and pre-diabetes in the population.


Link: https://bmcmedinformdecismak.biomedcentral.com/articles/10.1186/1472-6947-10-16
 

Strengths:

1. The real strength of SVM is it's ability to separate a non-linear data linearly using feature transformation which it does using Kernel Function i.e. there are some function that take low dimensional input space or feature space and map it to a very high dimensional space. Using this, a non-linearly separable data gets separated linearly.

2. In this we build a machine that only needs few support vectors i.e. the data points near to the decision boundary are considered. New input data is classified depending upon the similarity between the support vectors and the input data. We can say that SVM are fast in classifying new data provided we have appropriate number of support vector as compared to whole data. So, SVM perform well when we have few support vectors. As it uses a subset of training points in the decision function (called support vectors), it is also memory efficient.

3. Effective in high dimensional spaces.

Weakness:

1. SVM doesn't perform well in very large data sets because the training time happens to be cubic in size of the data sets.

2. SVM performs poorly with lots and lots of noise. So, when the class are overlapping we have to count independent evidence.

Good Candidate for problem:

The shape of our dataset is 395 datapoints and 48 features. Of which, training set contains 300 datapoints. Looking at features of datapoints, most features have categorical data(mostly binary data). So, SVM can be good candidate for our problem as we have small training set and categorical data in most features.

**2. K-Nearest Neighbors**

Real world Application:

General Application: Face Recognition, Pattern Recognition.

Classification of Heart Disease Using K- Nearest Neighbor and Genetic Algorithm

Link: http://research.ijcaonline.org/volume60/number14/pxc3881517.pdf

Strengths:

1. No wasting of time i.e. learning time is saved. It's simple.

2. As it does memorization, It doesn't forget. Thus, it's very reliable/dependable. If we give X again, we can get y back, instead of a smooth version of it (Using our approximated function). So, it remembers(No throwing away of information). Thus, it performs well in situations where exact output is required as opposed to it's smooth version. 

Weakness:

1. No generalization. Testing or predicting time is comparatively high.

2. It performs poorly when dataset shape is big. It takes too long to compute for new input data and also it ends up using lots of resources.

3. As it does memorization, it's going to overfit data. So, it's going to be sensitive to noise. It bottles the noise exactly, it bottles what it was given.

4. It performs poorly when we have two X's giving different y. Then lookup could return both and that could be problematic.

Good Candidate for problem:

Given the shape of our training dataset is small(300 datapoints and 48 features). And that most features have binary data, we can store it easily and also finding K-Nearest Neighbors won't take too long. Using K-Nearest Neighbors could be easy and efficient for our problem.

**3. Logistic Regression**

Real World Application

General Application: Any real world binary classification problem- win or lose, rain or not-rain, quit or not-quit etc.

Logistic Reggression was used to investigate employees turnover i.e. finding features and using them for predicting employees chances of quitting or not-quitting.
Doing this helped them make their interview process more focused on features that lead to quitting. Like in this case, commute distance.

Link: http://blog.minitab.com/blog/real-world-quality-improvement/using-binary-logistic-regression-to-investigate-high-employee-turnover 

Strengths

1. It's fast, less complex and easy to inspect.

2. It's computation cost is least when compared to other models.

Weakness 

1. It performs poorly with data having lots of noise.

2. It's prone to sampling bias. Thus, can sometimes suffer from overfitting.

Good Candidate for Problem

Logistic Regression is a simple algorithm and widely used for binary classification. It's one of the fastest algorithm used for binary classification as it's training time and testing time both are very low. In Logistic Regression, what happens mostly is that the coefficient of less important features are zero. Thus, features having non-zero coefficient value are considered when calculating probability using logistic function. So, as our problem being binary classification problem it seems to be good candidate for our problem. And also, it is fastest and efficient in performance.

### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [9]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))


### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [66]:
# TODO: Import the three supervised learning models from sklearn
# from sklearn import model_A
# from sklearn import model_B
# from skearln import model_C
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression

# TODO: Initialize the three models
clf_A = SVC()
clf_B = KNeighborsClassifier()
clf_C = LogisticRegression()

# TODO: Set up the training set sizes
X_train_100 = X_train[:100]
y_train_100 = y_train[:100]

X_train_200 = X_train[:200]
y_train_200 = y_train[:200]

X_train_300 = X_train[:300]
y_train_300 = y_train[:300]

# TODO: Execute the 'train_predict' function for each classifier and each training set size
# train_predict(clf, X_train, y_train, X_test, y_test)
train_predict(clf_A, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_A, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_A, X_train_300, y_train_300, X_test, y_test)
train_predict(clf_B, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_B, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_B, X_train_300, y_train_300, X_test, y_test)
train_predict(clf_C, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_C, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_C, X_train_300, y_train_300, X_test, y_test)

Training a SVC using a training set size of 100. . .
Trained model in 0.0040 seconds
Made predictions in 0.0040 seconds.
F1 score for training set: 0.8591.
Made predictions in 0.0040 seconds.
F1 score for test set: 0.8333.
Training a SVC using a training set size of 200. . .
Trained model in 0.0080 seconds
Made predictions in 0.0080 seconds.
F1 score for training set: 0.8581.
Made predictions in 0.0040 seconds.
F1 score for test set: 0.8408.
Training a SVC using a training set size of 300. . .
Trained model in 0.0080 seconds
Made predictions in 0.0080 seconds.
F1 score for training set: 0.8584.
Made predictions in 0.0040 seconds.
F1 score for test set: 0.8462.
Training a KNeighborsClassifier using a training set size of 100. . .
Trained model in 0.0000 seconds
Made predictions in 0.0040 seconds.
F1 score for training set: 0.7883.
Made predictions in 0.0000 seconds.
F1 score for test set: 0.7727.
Training a KNeighborsClassifier using a training set size of 200. . .
Trained model in 0.00

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1 - SVM**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |0.0040                   |0.0040                  |0.8591            |0.8333           |
| 200               |0.0080                   |0.0040                  |0.8581            |0.8408           |
| 300               |0.0080                   |0.0040                  |0.8584            |0.8462           |

** Classifer 2 - KNeighborClassifier**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |0.0000                   |0.0000                  |0.7883            |0.7727           |
| 200               |0.0000                   |0.0040                  |0.8345            |0.7971           |
| 300               |0.0000                   |0.0000                  |0.8558            |0.7681           |

** Classifer 3 - LogisticRegression**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |0.0000                   |0.0000                  |0.8529            |0.7727           |
| 200               |0.0000                   |0.0000                  |0.8269            |0.7857           |
| 300               |0.0000                   |0.0000                  |0.8337            |0.7368           |

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: **

I chose Logistic Regression as best model for our problem. It calculates predictive probability using Logistic/Sigmoid Function.
Input to Logistic Function is sum of product of co-efficient and value of feature.

As it assigns non-zero co-efficients to only those features which has Maximum Entropy after iterating through the data. It often happens in binary classification problem that features with non-zero co-efficient are very few at the end of final iteration i.e. features having non-zero co-efficients are the one on the value of which target label depends. And also simple and few resources are needed to store co-efficients and to perform simple mathematical operation of addition, multiplication and division used to perform above calculation for predicting label of input data. So, simple and few resources are needed and cost to calculate predictive probability is reduced. Thus, overall performance of model is increased. 

In KNeighbors, whole training set needs to be stored and prediction involves complex calculation. In SVM, support vectors needs to be stored i.e. subset of whole data and prediction involves complex calculation. Computation cost is high in SVM and KNeighbors. Large resources are needed to perform such complex calculations.

So, computation cost of Logistic Regression is least when compared to computation cost of SVM and KNeighbors.

Comparing training time and testing time of models above, we can see that Logistic Regression is the fastest of all three models. Though it has least F1 Score for training set of 300 samples. After tuning it's parameters, it can give best performance.

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: **

The best model chosen is Logistic Regression. In this, model is trained by finding features to be used for prediction. These are chosen depending upon information provided by that feature. Features providing maximum information are chosen for prediction. Information provided by feature is measured as sensitivity between target labels and value of that feature i.e. If value of feature is changed, there is change in target label or vice versa. Feature providing most information could be the one for which every change in it's value in data, there is change in target label. So, depending upon the measure of information provided, features are considered for predicting new input data.

The target label for new input data are predicted by taking in account the value of only those features which have be chosen for prediction. Remaining features of input data are simply ignored. This increases speed of model.

The model chosen is widely used for binary classification problem. Generally, this model is fatest, with minimum cost, using simple and least resources and has high metric score on unseen data. 

### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.GridSearchCV`](http://scikit-learn.org/0.17/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [78]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.metrics import make_scorer
from sklearn.grid_search import GridSearchCV

#TODO: Create the parameters list you wish to tune
parameters = {'max_iter':[9, 18, 27], 'penalty':['l1', 'l2'], 'random_state':[10, 50, 100], 'C':[0.001, 0.1, 1], 'n_jobs':[-1]}

#TODO: Initialize the classifier
clf = LogisticRegression()

# TODO: Make an f1 scoring function using 'make_scorer'
f1_scorer = make_scorer(f1_score, pos_label = 'yes')

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(clf, parameters, scoring = f1_scorer)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

#print clf
#print clf.coef_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

Made predictions in 0.0000 seconds.
Tuned model has a training F1 score of 0.8128.
Made predictions in 0.0000 seconds.
Tuned model has a testing F1 score of 0.8609.


## Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: **

The model's final F1 Score for training is 0.8128 and final F1 Score for testing is 0.8609. So, the model generalizes well to unseen data. Comparing it to untuned model, it can be seen that tuned model generalizes well to unssen data as compared to untuned model. Thus, we have huge increase of 0.1241 in F1 Score from untuned model to tuned model i.e. from 0.7368(untuned model) to 0.8609(tuned model).

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.